### PR TITLE
Integrating API for register route

### PR DESCRIPTION
**Pull request by pablonajera**
## Description
Integrating the API logic into the register route. 

### States
The Register route now handles:
- validation of email and password
- ensures password is "strong"
- allows user to view password
- shows an error message when the API returns an error
- show a success message the the API succeeds

### Components
Made many modifications to the `LabeledInput` component:
- `description` - can now optionally show a description below the label
- `type` - can now optionally handle different data types, including password inputs
  - options: number, text, password, maybe more
- `errorMessage` - can now optionally show an error message below the input
  -  If this is passed in, the text box also turns red to show it's in an error state
- `disabled` - optional boolean that grays out the textbox 
- `onBlur` - optional function to be run when the user clicks out of the input
- `icon` - optional icon to be shown to the right of the input
- `onIconClick` - optional function to be run whenever the user clicks on the icon.
  - **_Note_**: must pass in an icon first

## TODO
We still need to
- Link the "log in" button to the login page ([ticket](https://trello.com/c/5eNoIdsu))
- Link to home page when account successfully created ([ticket](https://trello.com/c/Gnu9P86p))
- Give more info on specific API errors, if possible. There seems to be a bug on API that is not allowing errors to be seen on UI side ([ticket](https://trello.com/c/BYTMNkjd))

## Screenshots
| Description | Screenshot   |
| -------  | ------- |
| Default |  Screenshot 2023-09-05 at 12 40 09 PM |
| Validation | Screenshot 2023-09-05 at 12 40 35 PM | 
| Loading |  Screenshot 2023-09-05 at 12 57 33 PM |
| Error | Screenshot 2023-09-05 at 12 57 43 PM |
| Success | Screenshot 2023-09-05 at 12 39 57 PM |

### CODE REVIEW

**Review by jnajera456 — APPROVED** [review]

**Review by angielamm — APPROVED** [review]

File: components/pieces/labeledInput/index.js
Comment (angielamm) [inline components/pieces/labeledInput/index.js:1]: Oooh I thought we were using a labeled input from Semantic UI but I'm glad we have our own! Could we use proptypes for this one as well?

File: constants/urls.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export const Urls = {
```
Comment (angielamm) [inline constants/urls.js:1]: Why did you add this url in a separate file and not in the register.js file?

File: pages/register.js [hunk 2]
```diff
@@ -1,72 +1,281 @@
...
+import _ from 'lodash-es';
+import validator from 'validator';
+import Link from 'next/link';
+
+export default function Register() {
+  // Create all the state variables we need to track the form inputs
+  const [name, setName] = useState('');
+  const [email, setEmail] = useState('');
+  const [password, setPassword] = useState('');
+  const [passwordConfirmation, setPasswordConfirmation] = useState('');
```
Comment (angielamm) [inline pages/register.js:25]: Just need to clarify:
Is using an empty string as default value of the state variables better than using null?

File: pages/register.js [hunk 3]
```diff
@@ -1,72 +1,281 @@
...
+  // It will be enabled if all the inputs are valid
+  const isSubmitButtonEnabled =
+    !_.isEmpty(name) &&
+    isEmailValid() &&
+    isPasswordValid() &&
+    isPasswordConfirmationValid();
+
+  // This will track the state of the API call
+  const { data, error, isLoading, trigger } = usePost({
+    url: `${Urls.API}/auth/signup`,
```
Comment (angielamm) [inline pages/register.js:46]: Why the $?

File: pages/register.js [hunk 4]
```diff
@@ -1,72 +1,281 @@
...
+    },
+  });
+
+  // useEffect will run the function passed in as the first argument
+  // every time the anything in the second argument changes
+  // in this case, we want to log the data and error every time they change
+  // in the future, we will want to do something with the data and error
+  useEffect(() => {
+    if (data) {
+      console.log(data);
```
Comment (angielamm) [inline pages/register.js:60]: This is only shown in the console right but not sent to the API address?

File: pages/register.js [hunk 5]
```diff
@@ -1,72 +1,281 @@
...
+    if (error) {
+      console.log(error);
+    }
+  }, [data, error]);
+
+  // This validates the email, password, and password confirmation
+  // every time they change
+  useEffect(() => {
+    if (isEmailValid()) {
+      setShowEmailError(false);
```
Comment (angielamm) [inline pages/register.js:72]: why false if it is initally set to false?

**Review by angielamm — COMMENTED** [review]
I've reviewed almost everything, I'm just missing going through the HTML. However, I've realized that the challenge lies in keeping track of all the functions that are interconnected

**Comment by pablonajera** [comment]
> I've realized that the challenge lies in keeping track of all the functions that are interconnected

Yup, that's a hard part of frontend!! I'll checkout your specific questions now.

File: components/pieces/labeledInput/index.js
Comment (pablonajera) [reply components/pieces/labeledInput/index.js:1]: Good idea 🧠 could you create a ticket and assign it to me?

File: constants/urls.js (hunk 1 above)
Comment (pablonajera) [reply constants/urls.js:1]: Good question! This URL will be used in many many many files, so it's better to define it in one place and just refer to it as a variable. You can see how it's used in the register page. 

In future integrations, you will use this variable as well!

File: pages/register.js (hunk 2 above)
Comment (pablonajera) [reply pages/register.js:25]: I'd say the only benefit is that the variable type is defined by using an empty string. When it's null, you don't even know what **_kind_** of variable it should be.

Besides that, not a huge difference

File: pages/register.js (hunk 3 above)
Comment (pablonajera) [reply pages/register.js:46]: Oh this is just string concatenation

```JavaScript
const name = 'Pablo'

console.log("Hello " + name)
// is the same as 
console.log(`Hello ${name}`)
```

We're just creating a full URL string using the base domain

File: pages/register.js (hunk 4 above)
Comment (pablonajera) [reply pages/register.js:60]: Right, for now I'm just logging the data to the console. In the future, we may save this info somewhere, etc.

File: pages/register.js (hunk 5 above)
Comment (pablonajera) [reply pages/register.js:72]: Because it might've been set to true. 

Starts as false --> set to true because there's some issue --> eventually, we need to set it back to false because there's no more error
